core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,710 @@
+
+    Returns:
+        dict. A dictionary containing tokens mapped to a unique ID.
+    """
+    # All unique tokens and number of time they occur in dataset. A token
+    # can be a keyword, an identifier, a constant, an operator.
+    vocabulary = {}
+
+    for pid in training_data:
+        program = training_data[pid]['source']
+        for token_id, token_name in get_tokens(program):
+            # Ignore all newline, comment and empty string tokens.
+            if (token_id == token.N_TOKENS or token_id == COMMENT_TOKEN
+                    or token_name.strip() == ''):
+                continue
+            # If token_id is tokens.NAME then only add if it is a python
+            # keyword.
+            elif token_id == token.NAME:
+                if token_name in keyword.kwlist:
+                    vocabulary[token_name] = vocabulary.get(token_name, 0) + 1
+            else:
+                vocabulary[token_name] = vocabulary.get(token_name, 0) + 1
+
+    # Consider only those tokens which occur for more than threshold times
+    # in entire dataset.
+    valid_tokens = [k for k, v in vocabulary.iteritems() if v > threshold]
+    token_to_id = dict(zip(valid_tokens, xrange(0, len(valid_tokens))))
+
+    # Add 'UNK' in token_to_id. This will be used to replace any token
+    # occurring in program which is not in valid_token.
+    token_to_id[UNK_TOKEN] = len(token_to_id)
+
+    # Add 'V' in token_to_id. This token will be used to replace all
+    # variables and methods in program.
+    token_to_id[VAR_TOKEN] = len(token_to_id)
+    return token_to_id
+
+
+def tokenize_data(training_data, threshold=VOCABULARY_THRESHOLD):
+    """Tokenize Python programs in dataset for winnowing.
+    Args:
+        training_data: dict. A dictionary containing training_data. Structure of
+            training_data is as follows,
+                {
+                    ID: {
+                        'source': str. Source code of program.
+                        'class': int. Associated feedback class with program
+                    } for each program in training_data.
+                }
+        threshold: int. Ignore a token if it appears less than 'threshold' times
+            in dataset.
+
+    Returns:
+        dict. A dictionary containing training data with additional key 'tokens'
+        for each data point which stores tokens of corresponding data point.
+        Structure of dict is as follows,
+            {
+                ID: {
+                    'source': str. Source code of program.
+                    'class': int. Associated feedback class with program.
+                    'tokens': list(str). Token list of program.
+                } for each program in training_data.
+            }
+    """
+    token_to_id = map_tokens_to_ids(training_data, threshold)
+
+    # Tokenize all programs in dataset.
+    for program_id in training_data:
+        program = training_data[program_id]['source']
+        token_program = []
```

Comment:
Call this token_names, or program_tokens, or tokens, since that's what it is (but see also the discussion on token vs token_name above). A token program sounds like a type of program.